libraries/SITL/SIM_JEDEC.cpp

```diff
@@ -0,0 +1,177 @@
+        storage_fd = open(filepath, flags, 0777);
+    }
+
+    // still -1 ??
+    if (storage_fd == -1) {
+        AP_HAL::panic("open(%s): %s", filepath, strerror(errno));
+    }
+    if (ftruncate(storage_fd, storage_size()) != 0) {
+        AP_HAL::panic("truncate(%s): %s", filepath, strerror(errno));
+    }
+}
+
+int JEDEC::rdwr(uint8_t count, SPI::spi_ioc_transfer *&tfrs)
+{
+    if (storage_fd == -1) {
+        open_storage_fd();
+    }
+
+    // commands:
+    static const uint8_t JEDEC_RDID             = 0x9f;
+    static const uint8_t JEDEC_READ             = 0x03;
+    static const uint8_t JEDEC_WREN             = 0x06;
+    static const uint8_t JEDEC_WRITE            = 0x02;
+    static const uint8_t JEDEC_RDSR             = 0x05;
+    static const uint8_t JEDEC_SECTOR4_ERASE    = 0x20;
+    static const uint8_t JEDEC_BULK_ERASE       = 0xC7;
+    static const uint8_t JEDEC_BLOCK64_ERASE    = 0xD8;
+
+    for (uint8_t i=0; i<count; i++) {
+        SPI::spi_ioc_transfer &tfr = tfrs[i];
+        uint8_t *tx_buf = (uint8_t*)(tfr.tx_buf);
+        uint8_t *rx_buf = (uint8_t*)(tfr.rx_buf);
+
+        switch (state) {
+        case State::WAITING: {
+            // find a command
+            uint8_t command = tx_buf[0];
+            switch (command) {
+            case JEDEC_RDID:
+                state = State::READING_RDID;
+                break;
+            case JEDEC_READ:
+                xfr_addr = tx_buf[1] << 16 | tx_buf[2] << 8 | tx_buf[3];
+                state = State::READING;
+                break;
+            case JEDEC_WRITE:
+                xfr_addr = tx_buf[1] << 16 | tx_buf[2] << 8 | tx_buf[3];
+                state = State::WRITING;
+                break;
+            case JEDEC_WREN:
+                write_enabled = true;
+                break;
+            case JEDEC_RDSR:
+                state = State::READING_RDSR;
+                break;
+            case JEDEC_SECTOR4_ERASE:   {
+                xfr_addr = tx_buf[1] << 16 | tx_buf[2] << 8 | tx_buf[3];
+                if (!write_enabled) {
+                    AP_HAL::panic("Writes not enabled");
+                }
+                if (xfr_addr + tfr.len > storage_size()) {
+                    abort();
+                }
+                uint8_t fill[get_page_size()*get_page_per_sector()];
+                memset(fill, 0xFF, sizeof(fill));
+                const size_t write_ret = pwrite(storage_fd, fill, sizeof(fill), xfr_addr);
+                if (write_ret != sizeof(fill)) {
+                    printf("Failed sector 4k erase");
+                }
+                write_enabled = false;
+                break;
+            }
+            case JEDEC_BULK_ERASE:  {
+                if (!write_enabled) {
+                    AP_HAL::panic("Writes not enabled");
+                }
+                if (xfr_addr + tfr.len > storage_size()) {
+                    abort();
+                }
+                uint8_t fill[storage_size()];
```

Comment:
Let's not put 32kB+ on the stack, please :-)

Loop instead, maybe a helper function for `SECTOR4_ERASE` and `BULK_ERASE`?

Reply:
Done! Have split bulk erase into multiple block erases and block erases into multiple sector erases.